## One hot encoding -variable with many categories 

In [8]:
#importing imp lib
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [10]:
data = pd.read_csv("train.csv", usecols=["X0","X1","X2","X3","X4","X5"])
data.head()

X0 X1  X2 X3 X4 X5
0   k  v  at  a  d  u
1   k  t  av  e  d  y
2  az  w   n  c  d  x
3  az  t   n  f  d  x
4  az  v   n  f  d  h

##### now checking how many unique labels each variable has

In [14]:
for col in data.columns:
    print(col+ " : " , len(data[col].unique()))

X0 :  47
X1 :  27
X2 :  44
X3 :  7
X4 :  4
X5 :  29


#### lets try np.get_dummies() fun

In [20]:
pd.get_dummies(data).shape

(4209, 158)

###### as we can see in this case using get dummies fun is not a good idea as we are getting 158 total variables earlier it was only 6

In [21]:
# so we will only take top 10 most frequent labels

In [39]:
data.X1.value_counts().sort_values(ascending=False).head(10)

aa    833
s     598
b     592
l     590
v     408
r     251
i     203
a     143
c     121
o      82
Name: X1, dtype: int64

In [34]:
top_10_X1 = [x for x in data.X1.value_counts().sort_values(ascending=False).head(10).index ]
top_10_X1

['aa', 's', 'b', 'l', 'v', 'r', 'i', 'a', 'c', 'o']

In [36]:
for label in top_10_X1:
    data[label] = np.where(data["X1"]==label,1,0)
    
data.head()

X0 X1  X2 X3 X4 X5  aa  s  b  l  v  r  i  a  c  o
0   k  v  at  a  d  u   0  0  0  0  1  0  0  0  0  0
1   k  t  av  e  d  y   0  0  0  0  0  0  0  0  0  0
2  az  w   n  c  d  x   0  0  0  0  0  0  0  0  0  0
3  az  t   n  f  d  x   0  0  0  0  0  0  0  0  0  0
4  az  v   n  f  d  h   0  0  0  0  1  0  0  0  0  0

### we will make a function in order to create dummy variables for each dataset

In [43]:
def one_head_encode(df , var):
    if len(data[var].value_counts() > 10):
        
        top_10_label = [x for x in data.X1.value_counts().sort_values(ascending=False).head(10).index]
    else:
        top_10_label = [x for x in data.X1.value_counts().sort_values(ascending=False).index]
    
    for label in top_10_label:
        
        data[var + "_" + label] = np.where(data[var]==label,1,0)
    
        
    

In [44]:
data = pd.read_csv("train.csv", usecols=["X0","X1","X2","X3","X4","X5"])

In [45]:
one_head_encode(data,"X1")
data.head()

X0 X1  X2 X3 X4 X5  X1_aa  X1_s  X1_b  X1_l  X1_v  X1_r  X1_i  X1_a  X1_c  \
0   k  v  at  a  d  u      0     0     0     0     1     0     0     0     0   
1   k  t  av  e  d  y      0     0     0     0     0     0     0     0     0   
2  az  w   n  c  d  x      0     0     0     0     0     0     0     0     0   
3  az  t   n  f  d  x      0     0     0     0     0     0     0     0     0   
4  az  v   n  f  d  h      0     0     0     0     1     0     0     0     0   

   X1_o  
0     0  
1     0  
2     0  
3     0  
4     0

In [51]:
one_head_encode(data,"X0")
one_head_encode(data,"X2")
one_head_encode(data,"X3")
one_head_encode(data,"X4")
one_head_encode(data,"X5")
data.shape

(4209, 66)

In [65]:
data.X1_aa.describe()*100

count    420900.000000
mean         19.790924
std          39.847070
min           0.000000
25%           0.000000
50%           0.000000
75%           0.000000
max         100.000000
Name: X1_aa, dtype: float64

#### from above we can now observe easily that there are 19.7% aa labels are present in variable x1

In [62]:
data.X1_s.describe()*100

count    420900.000000
mean         14.207650
std          34.917007
min           0.000000
25%           0.000000
50%           0.000000
75%           0.000000
max         100.000000
Name: X1_s, dtype: float64

In [63]:
data.X1_b.describe()*100

count    420900.000000
mean         14.065099
std          34.770247
min           0.000000
25%           0.000000
50%           0.000000
75%           0.000000
max         100.000000
Name: X1_b, dtype: float64

In [64]:
data.X1_v.describe()*100

count    420900.000000
mean          9.693514
std          29.590458
min           0.000000
25%           0.000000
50%           0.000000
75%           0.000000
max         100.000000
Name: X1_v, dtype: float64